src/NSSolverStationary.cpp

```diff
@@ -0,0 +1,933 @@
+              {
+                // compute both terms yielded by the Frechet derivative
+                // (u . nabla) u_old
+                nonlinear_term[k] += velocity_loc[q][l] *
+                                     fe_values[velocity].gradient(j, q)[k][l];
+
+                // (u_old . nabla) u
+                nonlinear_term[k] += fe_values[velocity].value(j, q)[l] *
+                                     velocity_gradient_loc[q][k][l];
+              }
+            }
+
+            // assemble the linearized convective term (u . nabla) uv
+            cell_matrix(i, j) += scalar_product(nonlinear_term, fe_values[velocity].value(i, q)) * fe_values.JxW(q);
+
+            // Third term - viscosity
+            cell_matrix(i, j) +=
+                nu *
+                scalar_product(fe_values[velocity].gradient(j, q),
+                               fe_values[velocity].gradient(i, q)) *
+                fe_values.JxW(q);
+
+            // Pressure term in the momentum equation.
+            cell_matrix(i, j) -= fe_values[pressure].value(j, q) *
+                                 fe_values[velocity].divergence(i, q) *
+                                 fe_values.JxW(q);
+
+            // Pressure term in the continuity equation.
+            cell_matrix(i, j) += fe_values[pressure].value(i, q) *
+                                 fe_values[velocity].divergence(j, q) *
+                                 fe_values.JxW(q);
+
+            // Pressure mass matrix used for preconditioning
+            cell_pressure_mass_matrix(i, j) +=
+                fe_values[pressure].value(i, q) *
+                fe_values[pressure].value(j, q) / nu * fe_values.JxW(q);
+          }
+        }
+
+        if (global_first_iter || computing_stokes)
+        {
+          continue;
+        }
+
+        //-R(u,v)
+        // a(u,v)
+        cell_rhs(i) -=
+            nu *
+            scalar_product(velocity_gradient_loc[q],
+                           fe_values[velocity].gradient(i, q)) *
+            fe_values.JxW(q);
+
+        // Compute residual associated with Newton linearization of (u . nabla) u
+        // nabla u is a tensor, iterate over both dimensions
+        for (unsigned int k = 0; k < dim; k++)
+        {
+          nonlinear_term[k] = 0.0;
+          for (unsigned int l = 0; l < dim; l++)
+          {
+            // compute (u_old . nabla) u_old
+            nonlinear_term[k] += velocity_loc[q][l] *
+                                 velocity_gradient_loc[q][k][l];
+          }
+        }
+
+        // assemble the residual associated with the nonlinear convective term
+        cell_rhs(i) -= scalar_product(nonlinear_term, fe_values[velocity].value(i, q)) * fe_values.JxW(q);
+
+        // b(v,p)
+        cell_rhs(i) += pressure_loc[q] *
+                       fe_values[velocity].divergence(i, q) *
+                       fe_values.JxW(q);
+
+        double velocity_divergence_loc = trace(velocity_gradient_loc[q]);
```

Comment:
```suggestion
        const double velocity_divergence_loc = trace(velocity_gradient_loc[q]);
```
Always marking const the variables that are const may help with compiler optimization.